### ISSUE TITLE
Attribute terms missing after 2.1.x -> 2.2.x update due to wc_sanitize_taxonomy_name() changes

### ISSUE THREAD

**Issue by om4james** [woothemes/woocommerce#6918]
After updating a WooCommerce store from 2.1.12 to 2.2.8, quite a few of variable products (when viewed on the frontend) showed an empty "Pa_food Packs" dropdown (instead of the proper attribute/variation dropdown), which meant the products couldn't be ordered.

Upon further investigation, many of the product attributes were showing as empty (ie with no terms in them).

The issue is caused by 83f123687540e8f3b169617e55d954a8cca5a884, which modifies `wc_sanitize_taxonomy_name ()` by introducing a newer (more strict) way of sanitising attribute names.

ie. WC 2.2.x+ more strictly sanitises the product attribute names, however it doesn't do any kind of database upgrade routine to convert the custom taxonomies to the new naming conventions.

The term/attribute data was in the database, but was still using the old `pa_` taxonomy name.

We resolved the issue by doing some search and replacing in the database:

```
wp search-replace 'pa_food-packs-' 'pa_food-packs'
wp search-replace 'pa_liner-widths-' 'pa_liner-widths'
wp search-replace 'pa_fish-sizes-' 'pa_fish-sizes'
wp search-replace 'pa_expanded-clay-' 'pa_expanded-clay'
wp search-replace 'pa_fish-nets---small' 'pa_pa_fish-nets-small'
wp search-replace 'pa_fish-nets---large' 'pa_pa_fish-nets-large'
wp search-replace 'pondmax-multi-use-pump-' 'pondmax-multi-use-pump'
```

We then went to Dashboard, Products, Attributes and edited each attribute, fixed slug to match the new naming conventions.

This is undoubtedly going to affect any WooCommerce store that uses product attributes that have extra spaces or characters in the name.

**Comment by mikejolley** [comment]
@om4james This has me a little perplexed.

If the attribute_names were sanitised differently, to pa_fish-nets---small for example, thats still a valid name. How would this break, unless the taxonomy was deleted then re-added (and thus re-sanitized)?

**Comment by om4james** [comment]
Hi @mikejolley!

On each page load, when register_taxonomy() is called for each product attribute, it uses the sanitised version of each product attribute:

https://github.com/woothemes/woocommerce/blob/1206e725045e5bf41603b7b10cbb6b3e80a7ed3a/includes/class-wc-post-types.php#L204

Line 164 calls wc_attribute_taxonomy_name() which in turn calls wc_sanitize_taxonomy_name(), which means that after upgrading WooCommmerce, then it could all of a sudden start looking for a new taxonomy name because of the new way each name is sanitised.

In other words, wc_sanitize_taxonomy_name() isn't just used when adding/updating attributes, it is also used on each and every page load.

Does that make sense?

James
